### ISSUE TITLE
hwloc-reader: config option to split resource "pools" (e.g., memory, storage)

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-sched#722]
Currently, the hwloc reader traverses the hwloc topology and creates a resource in the fluxion graph for every resources in the hwloc topology (subject to the allowlist). For "pool" resources like memory and node-local storage, the fluxion resource graph does not allow "shared exclusive" allocations of resource capacity (i.e., a job gets 2 GB of the capacity of a memory resource and another job gets the remaining memory capacity from the same resource). These two behaviors combined limit the capabilities of fluxion for certain workflows (e.g., memory-constrained ensembles).

One potential solution would be to have the hwloc reader automatically split resource "pools" based on some configurable unit of granularity (e.g., every 4 GBs).  This granularity could be hardcoded into a header file to start and then evolved to be read from the fluxion config file.

In #714, @dongahn suggested:
> In terms of storage resource vertex names, will storage[0]->chunk[k] work with where job spec is headed where storage is mapped to each block device and each chunk within it maps to a scheduleable unit pool? Or would flat storage_chunk[l] make more sense? I like the hierarchical specification as our pruning can take advantage of this for performance. How about memory? we can perhaps use memory[0]->chunk[m] similarly?

**Comment by grondo** [comment]
A naive question, sorry.

Are sparse jobspec requests still not supported? Would it have any other unintended side effects for jobspec creation and/or to introduce extra hierarchy by making chunks children of the actual resource?

From the core perspective (eg. generically querying, displaying, and changing status of resources), it would make much more sense if resource pool didn't introduce a "virtual" parent. (that is just my first impression)

**Comment by SteVwonder** [comment]
> Are sparse jobspec requests still not supported?  Would it have any other unintended side effects for jobspec creation and/or to introduce extra hierarchy by making chunks children of the actual resource?

Correct, they are still not supported.  So if we don't go the flat route, the user (or the front-end tool) will have to explicitly request the full hierarchy (e.g., `memory[1]->chunk[1:8g]`).  That'll get messy fast to handle portably across machines since it will depend on the "chunking" configuration and how hwloc/underlying resource file represents block devices.  For example, catalyst will require `storage[4]->chunk[3:64]` to get the entire node's SSD, while Lassen will require `storage[1]->storage[25:64]`.  Not so nice.

> it would make much more sense if resource pool didn't introduce a "virtual" parent. (that is just my first impression)

From a jobspec perspective, I think the flatter layout is easier to grok too.

**Comment by dongahn** [comment]
> Are sparse jobspec requests still not supported? Would it have any other unintended side effects for jobspec creation and/or to introduce extra hierarchy by making chunks children of the actual resource?

Maybe time to identify a few concrete job specs limited by this and investigate how much we can relax? I definitely cannot fit this into the end of Aug freeze though.

**Comment by SteVwonder** [comment]
> Maybe time to identify a few concrete job specs limited by this and investigate how much we can relax?

Sure.  Opened an issue to track that discussion: https://github.com/flux-framework/flux-sched/issues/727
